File: go/conformance_test.go
package hermetix

// 네 언어가 공유하는 골든 픽스처(conformance/fixtures)를 httptest 서버로 재생해 모든 어댑터를 컨포먼스 시나리오에 통과시킨다.

import (
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"os"
	"strings"
	"testing"
	"time"

	"github.com/shopspring/decimal"
)

type fixtureRoute struct {
	Method string          `json:"method"`
	Path   string          `json:"path"`
	Header []string        `json:"header"`
	Status int             `json:"status"`
	Body   json.RawMessage `json:"body"`
}

type fixtureFile struct {
	Scenario struct {
		Symbol     string `json:"symbol"`
		Quantity   string `json:"quantity"`
		LimitPrice string `json:"limitPrice"`
	} `json:"scenario"`
	Routes []fixtureRoute `json:"routes"`
}

func loadFixture(t *testing.T, broker string) (fixtureFile, *httptest.Server) {
	t.Helper()
	raw, err := os.ReadFile("../conformance/fixtures/" + broker + ".json")
	if err != nil {
		t.Fatal(err)
	}
	var fx fixtureFile
	if err := json.Unmarshal(raw, &fx); err != nil {
		t.Fatal(err)
	}
	// 픽스처 routes 를 순서대로 매칭 - method / path(정확히) / header(이름·값) 중 지정된 조건만 검사
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		for _, route := range fx.Routes {
			if route.Method != "" && route.Method != r.Method {
				continue
			}
			if route.Path != "" && route.Path != r.URL.Path {
				continue
			}
			if len(route.Header) == 2 && r.Header.Get(route.Header[0]) != route.Header[1] {
				continue
			}
			status := route.Status
			if status == 0 {
				status = 200
			}
			w.Header().Set("Content-Type", "application/json")
			w.WriteHeader(status)
			_, _ = w.Write(route.Body)
			return
		}
		w.WriteHeader(599)
		_, _ = w.Write([]byte(`{"error":"no fixture route for ` + r.Method + " " + r.URL.Path + `"}`))
	}))
	return fx, srv
}

func runConformance(t *testing.T, broker string, client func(baseURL string) BrokerClient) {
	t.Helper()
	fx, srv := loadFixture(t, broker)
	defer srv.Close()
	qty, _ := decimal.NewFromString(fx.Scenario.Quantity)
	limit, _ := decimal.NewFromString(fx.Scenario.LimitPrice)
	report := VerifyBrokerConformance(client(srv.URL), ConformanceScenario{Symbol: fx.Scenario.Symbol, Quantity: qty, LimitPrice: limit})
	if !report.Passed() {
		t.Fatal(report.String())
	}
	for _, s := range []string{"quotes", "candles", "calendar", "account", "holdings", "buyingPower", "createOrder", "getOrder", "getOrders", "cancelOrder", "fills"} {
		if !strings.Contains(strings.Join(report.Steps, ","), s) {
			t.Fatalf("step %s missing", s)
		}
	}
}

func TestNextConformance(t *testing.T) {
	runConformance(t, "next", func(baseURL string) BrokerClient {
		return NewNextClient("pk_test_conf", "sk_test_conf").SetBaseURL(baseURL)
	})
}

func TestKisConformance(t *testing.T) {
	runConformance(t, "kis", func(baseURL string) BrokerClient {
		return NewKisClient("k", "s", "50199202").SetBaseURL(baseURL).SetThrottle(time.Millisecond)
	})
}

func TestKiwoomConformance(t *testing.T) {
	runConformance(t, "kiwoom", func(baseURL string) BrokerClient {
		return NewKiwoomClient("k", "s").SetBaseURL(baseURL).SetThrottle(time.Millisecond)
	})
}

// accountNo 를 비워 /n2/acctinfo 로 모의(acct_type=03) 계좌를 고르는 경로까지 검증. 토큰은 운영 호스트 전용 — 테스트에선 같은 서버
func TestNhConformance(t *testing.T) {
	runConformance(t, "nh", func(baseURL string) BrokerClient {
		return NewNhClient("k", "s", "").SetBaseURL(baseURL).SetAuthURL(baseURL).SetThrottle(time.Millisecond)
	})
}

func TestDbConformance(t *testing.T) {
	runConformance(t, "db", func(baseURL string) BrokerClient {
		return NewDbClient("k", "s").SetBaseURL(baseURL).SetThrottle(time.Millisecond)
	})
}

func TestRateLimiterRetryAndThrottle(t *testing.T) {
	var sleeps []time.Duration
	now := time.Unix(0, 0)
	limiter := newRateLimiter(600*time.Millisecond, 2, func(a int) time.Duration { return time.Duration(a) * time.Second })
	limiter.sleep = func(d time.Duration) { sleeps = append(sleeps, d); now = now.Add(d) }
	limiter.throttle.sleep = limiter.sleep
	limiter.throttle.now = func() time.Time { return now }
	_, _ = limiter.execute("t", func() (map[string]any, error) { now = now.Add(100 * time.Millisecond); return nil, nil })
	_, _ = limiter.execute("t", func() (map[string]any, error) { return nil, nil })
	if len(sleeps) != 1 || sleeps[0] != 500*time.Millisecond {
		t.Fatalf("throttle sleeps = %v", sleeps)
	}

	sleeps = nil
	calls := 0
	limiter = newRateLimiter(0, 2, func(a int) time.Duration { return time.Duration(a) * time.Second })
	limiter.sleep = func(d time.Duration) { sleeps = append(sleeps, d) }
	body, err := limiter.execute("t", func() (map[string]any, error) {
		calls++
		if calls < 3 {
			return nil, newRateLimitError(429, "EGW00201", "초당 거래건수 초과")
		}
		return map[string]any{"ok": true}, nil
	})
	if err != nil || body["ok"] != true || len(sleeps) != 2 || sleeps[0] != time.Second || sleeps[1] != 2*time.Second {
		t.Fatalf("retry: err=%v sleeps=%v", err, sleeps)
	}

	sleeps = nil
	first := true
	limiter = newRateLimiter(0, 1, func(a int) time.Duration { return time.Second })
	limiter.sleep = func(d time.Duration) { sleeps = append(sleeps, d) }
	_, err = limiter.execute("t", func() (map[string]any, error) {
		if first {
			first = false
			return nil, newRateLimitErrorWithRetryAfter(429, "", "rl", 3600)
		}
		return nil, nil
	})
	if err != nil || len(sleeps) != 1 || sleeps[0] != maxRetryAfter {
		t.Fatalf("retry-after: err=%v sleeps=%v", err, sleeps)
	}

	_, err = limiter.execute("t", func() (map[string]any, error) { return nil, newRateLimitError(429, "", "x") })
	if _, ok := err.(*RateLimitError); !ok {
		t.Fatalf("소진 후 RateLimitError 전파: %v", err)
	}
}

func TestLsConformance(t *testing.T) {
	runConformance(t, "ls", func(baseURL string) BrokerClient {
		return NewLsClient("k", "s").SetBaseURL(baseURL).SetThrottle(time.Millisecond).SetChartThrottle(time.Millisecond)
	})
}

// accountSeq 를 비워 /api/v1/accounts 로 BROKERAGE 계좌를 고르는 경로까지 검증
func TestTossConformance(t *testing.T) {
	runConformance(t, "toss", func(baseURL string) BrokerClient {
		return NewTossClient("c_conf", "s_conf", "").SetBaseURL(baseURL).SetThrottle(time.Millisecond)
	})
}

func TestKbConformance(t *testing.T) {
	runConformance(t, "kb", func(baseURL string) BrokerClient {
		return NewKbClient("k", "s").SetBaseURL(baseURL).SetThrottle(time.Millisecond)
	})
}

File: js/tests/conformance.test.ts
/** 네 언어가 공유하는 골든 픽스처(conformance/fixtures)를 가짜 fetch 로 재생해 세 어댑터를 컨포먼스 시나리오에 통과시킨다. */
import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import { test } from "node:test";
import {
  Decimal, KisClient, KiwoomClient, NextClient, RateLimitError, RateLimiter, verifyBrokerConformance,
} from "../src/index.js";
import type { BrokerClient } from "../src/index.js";

interface Route { method?: string; path?: string; header?: [string, string]; status?: number; body: unknown; }

const fixturesDir = new URL("../../../conformance/fixtures/", import.meta.url);

function load(broker: string) {
  const fx = JSON.parse(readFileSync(new URL(`${broker}.json`, fixturesDir), "utf-8")) as {
    scenario: { symbol: string; quantity: string; limitPrice: string }; routes: Route[];
  };
  return {
    routes: fx.routes,
    scenario: { symbol: fx.scenario.symbol, quantity: new Decimal(fx.scenario.quantity), limitPrice: new Decimal(fx.scenario.limitPrice) },
  };
}

/** 픽스처 routes 를 순서대로 매칭 — method / path(정확히) / header(이름·값) 중 지정된 조건만 검사 */
function fakeFetch(routes: Route[]) {
  return async (url: string, init: { method: string; headers?: Record<string, string> }) => {
    const path = new URL(url).pathname;
    const headers = init.headers ?? {};
    const route = routes.find((r) =>
      (r.method === undefined || r.method === init.method) &&
      (r.path === undefined || r.path === path) &&
      (r.header === undefined || headers[r.header[0]] === r.header[1]));
    const status = route ? (route.status ?? 200) : 599;
    const body = route ? route.body : { error: `no fixture route for ${init.method} ${path}` };
    return { status, text: async () => JSON.stringify(body), headers: { forEach: () => {} } };
  };
}

async function run(broker: string, client: () => BrokerClient) {
  const { routes, scenario } = load(broker);
  const restore = (globalThis as any).fetch;
  (globalThis as any).fetch = fakeFetch(routes);
  try {
    const report = await verifyBrokerConformance(client(), scenario);
    assert.deepEqual(report.violations, [], report.violations.join("\n"));
    for (const s of ["quotes", "candles", "calendar", "account", "holdings", "buyingPower", "createOrder", "getOrder", "getOrders", "cancelOrder", "fills"]) {
      assert.ok(report.steps.includes(s), `step ${s} missing`);
    }
  } finally {
    (globalThis as any).fetch = restore;
  }
}

test("next 어댑터는 컨포먼스 시나리오를 통과한다", () => run("next", () => new NextClient("pk_test_conf", "sk_test_conf", "acc_main", "http://next.test")));
test("kis 어댑터는 컨포먼스 시나리오를 통과한다", () => run("kis", () => new KisClient("k", "s", "50199202", "01", "http://kis.test", 1)));
test("kiwoom 어댑터는 컨포먼스 시나리오를 통과한다", () => run("kiwoom", () => new KiwoomClient("k", "s", "http://kiwoom.test", 1)));

test("RateLimiter: 쓰로틀·백오프·Retry-After·재시도 소진", async () => {
  const sleeps: number[] = [];
  let now = 0;
  const sleeper = async (ms: number) => { sleeps.push(ms); now += ms; };
  const limiter = new RateLimiter(600, 2, (a) => 1000 * a, sleeper, () => now);
  await limiter.execute(async () => { now += 100; });
  await limiter.execute(async () => {});
  assert.deepEqual(sleeps, [500]);

  sleeps.length = 0;
  let calls = 0;
  const result = await new RateLimiter(0, 2, (a) => 1000 * a, sleeper).execute(async () => {
    calls++;
    if (calls < 3) throw new RateLimitError(429, "EGW00201", "초당 거래건수 초과");
    return "ok";
  });
  assert.equal(result, "ok");
  assert.deepEqual(sleeps, [1000, 2000]);

  sleeps.length = 0;
  let first = true;
  await new RateLimiter(0, 1, (a) => 1000 * a, sleeper).execute(async () => {
    if (first) { first = false; throw new RateLimitError(429, null, "rl", 3600); }
    return "ok";
  });
  assert.deepEqual(sleeps, [RateLimiter.MAX_RETRY_AFTER_MS]);

  await assert.rejects(new RateLimiter(0, 1, () => 0, sleeper).execute(async () => { throw new RateLimitError(429, null, "x"); }), RateLimitError);
});
